### PR TITLE
[NO SQUASH] Meshgen tests

### PR DESCRIPTION
**Pull request by numberZero**
Add compact, short information about your PR for easier understanding:

- Goal of the PR
  - Meshgen testability.
- How does the PR work?
  1. Decouple meshgen from `Client`, to allow testing
  2. Add mesh comparison functions
  3. Add some basic meshgen tests using those
- Does it resolve any reported issue?
  - No.
- Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?
  - Yes, 2.2 Internal code refactoring
- If not a bug fix, why is this PR needed? What usecases does it solve?

## To do

This PR is Ready for Review

## How to test

* Run Minetest (any game), make sure the nodes are still rendered
* Run unit tests

### CODE REVIEW

**Comment by numberZero** [comment]
Take a look at non-test code changes here, please. Some changes are necessary for testability (since full `Client` is not mockable at all) but there may be other ways.

File: src/client/mesh_generator_thread.cpp [hunk 1]
```diff
@@ -227,7 +228,7 @@ void MeshUpdateWorkerThread::doUpdate()
 			sleep_ms(m_generation_interval);
 		ScopeProfiler sp(g_profiler, "Client: Mesh making (sum)");
 
-		MapBlockMesh *mesh_new = new MapBlockMesh(q->data, *m_camera_offset);
+		MapBlockMesh *mesh_new = new MapBlockMesh(m_client, q->data, *m_camera_offset);
```
Comment (sfan5) [inline src/client/mesh_generator_thread.cpp:231]: could be `m_manager->m_client` (classes are tightly coupled anyway)

File: src/main.cpp [hunk 2]
```diff
@@ -217,7 +217,10 @@ int main(int argc, char *argv[])
 	// Run unit tests
 	if (cmd_args.getFlag("run-unittests")) {
 #if BUILD_UNITTESTS
-		return run_tests();
+		if (cmd_args.exists("test-module"))
+			return run_tests(cmd_args.get("test-module")) ? 0 : 1;
```
Comment (sfan5) [inline src/main.cpp:221]: :+1: I was missing something like this the other day!

**Review by sfan5 — COMMENTED** [review]
lgtm

**Comment by Zughy** [comment]
@numberZero rebase needed

**Comment by numberZero** [comment]
> @numberZero rebase needed

For a draft?

**Comment by Zughy** [comment]
I thought it was kind of ready due to the sfan5's "lgtm", my bad

**Comment by numberZero** [comment]
Ready for review.

There are only a few tests included; more should follow in other PRs. E.g. for facedir, raillike, connected nodeboxes...

**Review by sfan5 — APPROVED** [review]

File: src/unittest/test_mesh_compare.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+public:
+	TestMeshCompare() { TestManager::registerTestModule(this); }
+	const char *getName() override { return "TestMeshCompare"; }
+
+	void runTests(IGameDef *gamedef) override {
+		TEST(testTriangle);
+		TEST(testQuad);
+	}
+
+	void testTriangle() {
```
Comment (SmallJoker) [inline src/unittest/test_mesh_compare.cpp:34]: What is the purpose of these tests? What could *possibly* go wrong with triangles?

Comment (numberZero) [reply src/unittest/test_mesh_compare.cpp:34]: Triangles are fine. Comparing them however...
And this `TestMeshCompare::testTriangle` tests that comparison of triangle-based meshes works as intended. Or at least it should; apparently I forgot to add cases with more than one triangle.

Comment (SmallJoker) [reply src/unittest/test_mesh_compare.cpp:34]: So if I understand you correctly, is this to check whether the comparisons within the tests themselves are correct? Basically a test for the unittest?

Comment (numberZero) [reply src/unittest/test_mesh_compare.cpp:34]: Yes. Anything wrong with that?

**Review by SmallJoker — APPROVED** [review]
Fine by me. The added unittests pass.

Added a short description for test_mesh_compare.cpp to understand what it does and squashed the commit into "Add mesh comparison functions, for tests".

**Pull request merged**